File: lib/src/main/java/com/hammwerk/wizardpager/core/BranchPage.java
package com.hammwerk.wizardpager.core;

import java.util.ArrayList;
import java.util.List;

public abstract class BranchPage extends Page<Integer> {
	protected List<Branch> branches = new ArrayList<>();
	private Branch selectedBranch;
	private BranchPageListener branchPageListener;

	public BranchPage(String title) {
		super(title, true);
	}

	public BranchPage addBranch(String branchName, Page... pages) {
		branches.add(new Branch(branchName)
				.setPages(pages));
		return this;
	}

	public void selectBranch(int index) {
		if (branches.size() < 2) {
			throw new TwoBranchesRequiredException();
		}
		if (!branches.get(index).equals(selectedBranch)) {
			selectedBranch = branches.get(index);
			setResult(index);
			setCompleted();
			if (branchPageListener != null) {
				branchPageListener.onBranchSelected(this);
			}
		}
	}

	public Branch getSelectedBranch() {
		if (branches.size() < 2) {
			throw new TwoBranchesRequiredException();
		}
		return selectedBranch;
	}

	public String[] getChoices() {
		if (branches.size() < 2) {
			throw new TwoBranchesRequiredException();
		}
		String[] choices = new String[branches.size()];
		for (int i = 0; i < branches.size(); i++) {
			choices[i] = branches.get(i).getName();
		}
		return choices;
	}

	void setBranchPageListener(BranchPageListener listener) {
		this.branchPageListener = listener;
	}

	public class TwoBranchesRequiredException extends RuntimeException {
	}
}

File: lib/src/main/kotlin/com/hammwerk/wizardpager/core/BranchPage.kt
package com.hammwerk.wizardpager.core

open class BranchPage(title: String, fragmentName: String) :
		Page(title, fragmentName, true) {
	protected var branches = mutableListOf<Branch>()
	var onBranchSelected: ((BranchPage) -> Unit)? = null

	val choices: Array<String?>
		get() {
			if (branches.size < 2) {
				throw TwoBranchesRequiredException()
			}
			return branches.map { it.name }.toTypedArray()
		}

	val selectedBranchIndex: Int?
		get() {
			val selectedBranchIndex = branches.indexOf(selectedBranch)
			return when {
				selectedBranchIndex != -1 -> selectedBranchIndex
				else -> null
			}
		}

	var selectedBranch: Branch? = null
		get() = when {
			branches.size >= 2 -> field
			else -> throw TwoBranchesRequiredException()
		}
		private set

	fun addBranch(branchName: String, vararg pages: Page) {
		branches.add(Branch(branchName).apply { this.pages = arrayOf(*pages) })
	}

	fun selectBranch(index: Int) {
		if (branches.size < 2) {
			throw TwoBranchesRequiredException()
		}
		if (branches[index] != selectedBranch) {
			selectedBranch = branches[index]
			completed = true
			onBranchSelected?.let { it(this) }
		}
	}

	class TwoBranchesRequiredException : RuntimeException()
}
